File: Restaurant-information-system/src/main/java/com/galaxy/Restaurantinformationsystem/Service/ReviewService.java
package com.galaxy.Restaurantinformationsystem.Service;

import com.galaxy.Restaurantinformationsystem.DTO.ReviewDTO;
import com.galaxy.Restaurantinformationsystem.DTO.ReviewFormDTO;
import com.galaxy.Restaurantinformationsystem.Entity.ReviewEntity;
import com.galaxy.Restaurantinformationsystem.Entity.StoreEntity;
import com.galaxy.Restaurantinformationsystem.Repository.MenuRepository;
import com.galaxy.Restaurantinformationsystem.Repository.ReviewRepository;
import com.galaxy.Restaurantinformationsystem.Repository.StoreRepository;
import com.galaxy.Restaurantinformationsystem.Repository.UserRepository;
import lombok.AllArgsConstructor;
import org.modelmapper.ModelMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class ReviewService {

    ReviewRepository reviewRepository;
    MenuRepository menuRepository;
    UserRepository userRepository;
    StoreRepository storeRepository;
    ModelMapper modelMapper;


    public ReviewDTO createReview(Long storeId, ReviewFormDTO reviewFormDTO, String token) {
        ReviewEntity reviewEntity = modelMapper.map(reviewFormDTO, ReviewEntity.class);
        reviewEntity.setStoreEntity(StoreEntity.builder().id(storeId).build());
        // TODO: token으로 부터 id를 받아와서 userEntity에 넣어주기
        //reviewEntity.setUserEntity(UserEntity.builder().id().build());
        return modelMapper.map(reviewRepository.save(reviewEntity), ReviewDTO.class);
    }

    public ReviewDTO updateReview(Long reviewid, ReviewFormDTO reviewFormDTO, String token) {
        Optional<ReviewEntity> reviewEntity = reviewRepository.findById(reviewid);

        if (reviewEntity.isEmpty()) {
            throw new RuntimeException("리뷰가 존재하지 않습니다.");
        }
//         작성자 본인이 아니면 수정 불가
//        if (reviewEntity.get().getUserEntity().getId() != 1L){
//            throw new RuntimeException("권한이 없습니다.");
//        }

        reviewEntity.get().setTitle(reviewFormDTO.getTitle());
        reviewEntity.get().setContent(reviewFormDTO.getContent());
        reviewEntity.get().setImage(reviewFormDTO.getImage());
        reviewEntity.get().setScore(reviewFormDTO.getScore());
        return modelMapper.map(reviewRepository.save(reviewEntity.get()), ReviewDTO.class);

    }

    public List<ReviewDTO> getMyReview(String token) {
        // token에서 id 받아오기
        Long userId = 1L;
        List<ReviewEntity> reviewEntityList = reviewRepository.findByUserEntity_Id(userId);

        List<ReviewDTO> reviewDTOList = new ArrayList<>();
        for (ReviewEntity reviewEntity : reviewEntityList) {
            reviewDTOList.add(modelMapper.map(reviewEntity, ReviewDTO.class));
        }

        return reviewDTOList;
    }

    @Transactional
    public void deleteReview(Long reviewId, String token) {
        Optional<ReviewEntity> reviewEntity = reviewRepository.findById(reviewId);

        if (reviewEntity.isEmpty()) {
            throw new RuntimeException("리뷰가 존재하지 않습니다.");
        }

        // 작성자 본인이 아니면 삭제 불가
//        if (reviewEntity.get().getUserEntity().getId() != 1L){
//            throw new RuntimeException("권한이 없습니다.");
//        }

        reviewRepository.delete(reviewEntity.get());
    }

}

File: src/main/java/com/galaxy/restaurant/information/system/restaurant/service/ReviewService.kt
package com.galaxy.restaurant.information.system.restaurant.service

import com.galaxy.restaurant.information.system.restaurant.controller.request.ReviewCreateForm
import com.galaxy.restaurant.information.system.restaurant.controller.response.ReviewResponse
import com.galaxy.restaurant.information.system.restaurant.entity.Review
import com.galaxy.restaurant.information.system.restaurant.repository.RestaurantRepository
import com.galaxy.restaurant.information.system.restaurant.repository.ReviewRepository
import com.galaxy.restaurant.information.system.user.repository.UserRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ReviewService(
    private var reviewRepository: ReviewRepository,
    private var restaurantRepository: RestaurantRepository,
    private var userRepository: UserRepository,
) {
    /**
     * 리뷰를 생성합니다.
     * @param storeId 가게 id
     * @param reviewFormDTO 리뷰 생성 요청 DTO
     * @param username 사용자 이름
     * @return 생성된 리뷰 정보
     */
    @Transactional
    fun createReview(storeId: Long, reviewFormDTO: ReviewCreateForm, username: String): ReviewResponse {
        val review = Review(
            title = reviewFormDTO.title,
            content = reviewFormDTO.content,
            image = reviewFormDTO.image,
            score = reviewFormDTO.score,
            store = restaurantRepository.findById(storeId).get(),
            author = userRepository.findByEmail(username) ?: throw RuntimeException("가입되지 않은 유저입니다.")
        )
        return storeMapper.toResponseDto(reviewRepository.save(review))
    }

    /**
     * 리뷰를 수정합니다.
     * @param reviewid 리뷰 id
     * @param reviewFormDTO 리뷰 수정 요청 DTO
     * @param username 사용자 이름
     * @return 수정된 리뷰 정보
     */
    @Transactional
    fun updateReview(reviewid: Long, reviewFormDTO: ReviewCreateForm, username: String): ReviewResponse {
        val reviewEntity = reviewRepository.findById(reviewid).orElseThrow()
        reviewEntity.title = reviewFormDTO.title
        reviewEntity.content = reviewFormDTO.content
        reviewEntity.image = reviewFormDTO.image
        reviewEntity.score = reviewFormDTO.score
        return storeMapper.toResponseDto(reviewRepository.save(reviewEntity))
    }

    /**
     * 리뷰를 삭제합니다.
     * @param reviewId 리뷰 id
     */
    @Transactional
    fun deleteReview(reviewId: Long) {
        reviewRepository.deleteById(reviewId)
    }

    @Transactional(readOnly = true)
    fun getUserReviews(userId: Long, pageable: Pageable): Page<ReviewResponse> {
        return reviewRepository.findByAuthorId(userId, pageable).map { review ->
            ReviewResponse.from(review)
        }
    }
}
